src/client/Client.cc

```diff
@@ -5828,7 +5828,11 @@ int Client::may_setattr(Inode *in, struct ceph_statx *stx, int mask,
   }
 
   if (mask & CEPH_SETATTR_MODE) {
-    if (perms.uid() != 0 && perms.uid() != in->uid)
+    uint32_t m = ~stx->stx_mode & in->mode;
+    if (perms.uid() != 0 && perms.uid() != in->uid &&
+	// Only when any of S_ISUID and S_ISUID mode bits is cleared
```

Comment:
you probably meant to say **`S_ISUID` and `S_ISGID`**

also, my head hurts trying to wrap my head around the double negation in the condition below

Reply:
Yeah, correct. I have adjust this comments in new commits below. Thanks @mchangir